src/data.rs

```diff
@@ -18,101 +18,162 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
-use config::Config;
-use csv;
-use prelude::*;
-use serde_json;
-use std::collections::HashMap;
+use enum_map::{Enum, EnumMap};
+use serde_derive::{Deserialize, Serialize};
+use tokio::sync::Mutex;
+use tokio::task::{spawn_blocking, JoinSet};
+
+use crate::config::Config;
+use std::collections::BTreeMap;
+use std::fs::OpenOptions;
 use std::path::PathBuf;
-use std::sync::{Arc, Mutex};
+use std::sync::atomic::AtomicUsize;
+use std::sync::Arc;
 use std::{
-    fs::{self, File, OpenOptions},
+    fs::{self, File},
     io::{prelude::*, BufWriter},
 };
 
-#[derive(Default, Serialize, Deserialize)]
-struct State {
-    last_id: HashMap<String, usize>,
+#[derive(Debug, Enum, Serialize, Deserialize, Copy, Clone)]
+pub enum Forge {
+    Github,
 }
 
-#[derive(Serialize, Deserialize)]
+#[derive(Debug, Default, Serialize, Deserialize)]
+struct State(EnumMap<Forge, AtomicUsize>);
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
 pub struct Repo {
     pub id: String,
     pub name: String,
     pub has_cargo_toml: bool,
     pub has_cargo_lock: bool,
 }
 
+#[derive(Debug, Clone)]
```

Comment:
Could a single Arc for `self` be more efficient? Possibly with separate mutexes still?